docs/design/auth-layer.md

```diff
@@ -0,0 +1,224 @@
+2. Parse as a valid JSON-RPC request.
+3. Extract the access token from the header.
+4. Extract the "method" and "params" from the JSON-RPC request.
+5. Validate the access token with the authentication server and get the user's permissions.
+6. If the user has the required permissions, forward the request to the Hedera-The-Graph node.
+7. If the user does not have the required permissions, return an error (401 Unauthorized).
+
+For the implementation of the `Auth-Proxy` we will use a Envoy Proxy with a custom HTTP filter using LUA Script to validate the access token, and associated permissions.
+
+Environtment variables to configure the `Auth-Proxy`:
+
+- **TOKEN_INTROSPECTION_URL:** needed to validate the access token with the authentication server and retrieve token claims.
+- **CLIENT_ID:** needed to identify the application that is verifying the token veracity and claims.
+- **CLIENT_SECRET:** needed to authenticate the application that is verifying the token veracity and claims.
+
+
+### C. Authentication Server
+
+This component is responsible for authenticating users and providing access tokens. The authentication server is responsible for validating the user's credentials and generating an access token. The access token is used to authenticate the user when making requests to the Hedera-The-Graph node. The authentication server will use OAuth2 and JWT standards to provide access tokens, and it will manage the custom claims to define the user's permissions. 
+
+Any authentication server that supports OAuth2 and JWT standards can be used, but for the implementation included on this project we will use a `Keycloak` server, it was chosen because it is a robust OSS Identity Provider that is widely used and has a lot of documentation and community support available and it is easy to integrate with other systems and to manage users, roles, and permissions.
+
+We will be using the following endpoints of the Keycloak server:
+- **Token:** `/realms/HederaTheGraph/protocol/openid-connect/token` to obtain the access token.
+- **Token Introspect:** `/realms/HederaTheGraph/protocol/openid-connect/token/introspect` to validate the access token and retrieve token claims.
+
+Users permissions will be defined with 3 custom roles: "subgraph_create", "subgraph_deploy" and "subgraph_remove". And with a custom claim called "subgraph_access" that will contain a CSV String with the subgraphs names that the user has access to.
+
+#### Roles: 
+1. **subgraph_create:** allows the user to create a new subgraph.
+2. **subgraph_deploy:** allows the user to deploy a subgraph.
+3. **subgraph_remove:** allows the user to remove a subgraph.
+
+#### Custom Claims:
+- **subgraph_access:** contains a CSV String with the subgraphs names that the user has access to. ie: "subgraph1,subgraph2,subgraph3"
+
+#### Token Request example:
+
+```bash
+curl --location 'http://host.docker.internal:8080/realms/HederaTheGraph/protocol/openid-connect/token' \
```

Comment:
There's a few fields in this call that need to exist first.
You should call out what expected setup has already performed e.g. a user profile is already setup with x y and Z configs

Reply:
✅ Done